google/cloud/spanner/snapshot.py

```diff
@@ -150,17 +152,18 @@ def read(self, table, columns, keyset, index="", limit=0, partition=None):
         metadata = _metadata_with_prefix(database.name)
         transaction = self._make_txn_selector()
 
-        restart = functools.partial(
-            api.streaming_read,
-            self._session.name,
-            table,
-            columns,
-            keyset._to_pb(),
+        request = ReadRequest(
+            session=self._session.name,
+            table=table,
+            columns=columns,
+            key_set=keyset._to_pb(),
```

Comment:
Is this conversion still necessary?

Reply:
Yes. `keyset` is a handwritten wrapper for the `KeySet` proto. Arguably, with the `proto-plus` generated types, this handwritten wrapper could be removed and the `KeySet` proto could be used directly. My only concern about doing this is whether the error messages from the proto for invalid inputs would be as clear and helpful.

Reply:
I did a trial refactor where the handwritten `KeySet` and `KeyRange` were removed and ran into some issues with type conversion due to Spanner expecting a `string_value` for `INT64` values. So at this stage, I think keeping these handwritten wrappers makes sense.